File: GW3/LogFileHandling/Query/Tokens/TokenOrder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GWLogger.Backend.DataContext.Query.Tokens
{
    internal class TokenOrder : Token
    {
        public override bool CanBeUsed(QueryParser parser)
        {
            parser.SkipSpaces();

            return parser.PeekString().ToLower() == "order";
        }

        public override Token Extract(QueryParser parser)
        {
            parser.SkipSpaces();
            parser.NextString();
            parser.SkipSpaces();
            if (parser.PeekString().ToLower() == "by")
                parser.NextString();
            return new TokenOrder();
        }
    }
}

File: GW3/GWLogger/Frontend/Query/Tokens/TokenOrder.ts
class TokenOrder extends Token {

    public tokenType: TokenType = TokenType.TokenOrder;

    public proposals: SuggestionInterface[] = <SuggestionInterface[]>[{ suggestion: "order by" }];

    public canBeUsed(parser: QueryParser): boolean {
        parser.skipSpaces();
        return parser.peekString().toLowerCase() == "order";
    }
    public extract(parser: QueryParser): Token {
        parser.skipSpaces();
        parser.nextString();
        parser.skipSpaces();
        let token = new TokenOrder();
        token.value = "limit";
        if (parser.peekString().toLowerCase() == "by") {
            token.value += " " + parser.nextString();
        }
        return token;
    }

    public getProposals(nextToken?: Token, afterNextToken?: Token): Token[] {
        if (typeof nextToken != 'undefined' && typeof afterNextToken != 'undefined') {
            switch (nextToken.tokenType) {
                case TokenType.TokenName:
                    return [
                        new TokenAscending(),
                        new TokenDescending(),
                        new TokenGroup(),
                        new TokenLimit(),
                        new TokenOrder(),
                        new TokenWhere()
                    ];
            }
        }
        return [];
    }
}
